docs/src/dev/principles/unsafe.md

```diff
@@ -0,0 +1,270 @@
+the function cannot guarantee the safety of the unsafe code blocks within it without a contract with the caller should
+a function be marked as unsafe.
+
+For software safety, all of the above holds true. If software must guarantee a pre/postcondition or an invariant in
+order to safely use a function, e.g.
+
+```rust
+/// Returns a reference to the element at the specified index without performing bounds checking.
+///
+/// # Parameters
+/// - `index`: The position of the element to retrieve.
+///
+/// # Returns
+/// A reference to the element at the given index.
+///
+/// # Safety
+/// Calling this function with an out-of-bounds index is undefined behavior.
+/// The caller must ensure that `index` is within the bounds of the collection.
+unsafe fn get_element_unchecked<T>(slice: &[T], index: usize) -> &T {
+  // SAFETY: Caller must ensure that index < slice.len()
+  unsafe { slice.get_unchecked(index) }
+}
+```
+
+In this case, the invariant is that `index` is within the bounds of `slice`. It is up to the programmer to decide how
+this interface should be defined and whether the function is unsafe. For example, the function could have been written
+as:
+
+```rust
+/// Returns an Option containing a reference to the element at the specified index without performing bounds checking.
+///
+/// # Parameters
+/// - `index`: The position of the element to retrieve.
+///
+/// # Returns
+/// An Option containing a reference to the element at the given index.
+///
+/// # Safety
+/// Calling this function with an out-of-bounds index is undefined behavior.
+/// The caller must ensure that `index` is within the bounds of the collection.
+fn get_element_unchecked<T>(slice: &[T], index: usize) -> Option<&T> {
+  if index >= slice.len() {
+    return None;
+  }
+  // SAFETY: Caller must ensure that index < slice.len()
+  unsafe { Some(slice.get_unchecked(index)) }
+}
+```
+
+In this version, the function is no longer unsafe, despite containing an unsafe code block within it, because it no
+longer has an invariant; the function itself is able to manage whether its inputs are valid.
+
+Whenever possible, programmers should write safe functions that validate all preconditions, postconditions, and
+invariants. Of course, this is not always possible, in which case unsafe functions are called for.
+
+For hardware safety, we don't have invariants, necessarily. For example, in writing a system register, there is no
+invariant that must hold true for that operation. There may be preconditions or postconditions, but not an invariant.
+
+One simple example of hardware safety is the invalidate cache instruction on x86_64 that invalidates caches without
+writing them back. This operation can be very dangerous if not called at a proper time, but nothing about the software
+can change that; there is no validation that can be done to indicate this is a safe time to call this. In Patina, this
+is hardware safety and does not warrant an unsafe function.

```

Comment:
> there is no validation that can be done to indicate this is a safe time to call this. In Patina, this is hardware safety and does not warrant an unsafe function.

I would hold back from such a strong statement. There are cases like executing in Cache-As-RAM that you can check the NEM state might choose to prevent executing `invd` to prevent flushing all of the verified code content and NEM data in the CPU cache. That doesn't occur in DXE, but I also don't think that distinction matters here.

The description doesn't need to be super pedantic but I'd avoid saying "there is no validation that can be done".